### PR TITLE
[SPARK-47107][SS][PYTHON] Implement partition reader for python streaming data source

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Piggy back the PythonPartitionReaderFactory to implement reading a data partition for python streaming data source. Add test case to verify that python streaming data source can read and process data end to end.

### Why are the changes needed?
This is part of the effort to support developing streaming data source in python interface.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add integration test to verify data are read and metrics are emitted correctly.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by chaoqin-li1123** [comment]
@HeartSaVioR @HyukjinKwon @allisonwang-db PTAL, thanks!

**Review by HyukjinKwon — APPROVED** [review]
Seems fine from a cursory look.

**Review by HeartSaVioR — COMMENTED** [review]
Mostly looks great. Would be it feasible to add python e2e test for this?

**Review by allisonwang-db — APPROVED** [review]

File: python/pyspark/sql/worker/plan_data_source_read.py [hunk 1]
```diff
@@ -264,11 +246,40 @@ def batched(iterator: Iterator, n: int) -> Iterator:
...
+                partitions = [None]
+
+            # Return the serialized partition values.
+            write_int(len(partitions), outfile)
+            for partition in partitions:
+                pickleSer._write_with_length(partition, outfile)
+        else:
+            # Send an empty list of partition for stream reader because partitions are planned
+            # in each microbatch during query execution.
+            write_int(0, outfile)
```
Comment (allisonwang-db) [inline python/pyspark/sql/worker/plan_data_source_read.py:282]: Could you give me some code pointers on where the planning of partitions takes place in streaming?

Comment (chaoqin-li1123) [reply python/pyspark/sql/worker/plan_data_source_read.py:282]: It is in PythonMicrobatchStream's planInputPartitions() https://github.com/apache/spark/blob/6a27789ad7d59cd133653a49be0bb49729542abe/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonMicroBatchStream.scala#L48

Comment (allisonwang-db) [reply python/pyspark/sql/worker/plan_data_source_read.py:282]: Got it thanks!

**Comment by HeartSaVioR** [comment]
@chaoqin-li1123 Could you please rebase this? Looks like there are conflicts now.

**Review by HeartSaVioR — APPROVED** [review]
+1 pending CI.

**Comment by HeartSaVioR** [comment]
@chaoqin-li1123 Could you please file a JIRA ticket to add end-to-end python test suite for both streaming read and streaming write? I'm not confident that ScalaTest based test is sufficient.

**Comment by HeartSaVioR** [comment]
In the meanwhile, I'm merging to master as CI has passed.

**Pull request closed without merging**